### PR TITLE
Fix memory leak in bc.apply + cleanups

### PR DESCRIPTION
**Pull request by wence-**
The `bc.apply` annotation unconditionally saved the bc on the function it was applied to. Causing a memory leak. This was noticed in defcon where new bc objects are made as part of the continuation/deflation process, but applied to the same state vector.

This is clearly a memory leak full stop, but I don't have any pyadjoint-fu, so I'm not going to try and fix it there. At the very least, this patch removes the behaviour that in annotations off mode, something still happens.

In addition, a bunch of cleanups such that if we're not in adjoint mode, the wrappers/decorators really are no-ops.

Please audit carefully.

### CODE REVIEW

**Comment by pefarrell** [comment]
The test that this fails is `FAILED tests/regression/test_adjoint_bc.py::test_bcs_collected_when_not_annotating`.

This seems counterintuitive. Why do we test that the BCs are collected even when not anotating?

**Comment by wence-** [comment]
> Why do we test that the BCs are collected even when not anotating?

collected in the sense of garbage collected.

**Comment by pefarrell** [comment]
Apologies, I misunderstood.

File: .gitignore [hunk 1]
```diff
@@ -9,12 +9,14 @@
 *.pvtu
 *.so
 *.vtu
+.DS_Store
```
Comment (pefarrell) [inline .gitignore:12]: This doesn't seem like a relevant change?

File: firedrake/adjoint/dirichletbc.py [hunk 2]
```diff
@@ -20,14 +20,17 @@ def wrapper(self, *args, **kwargs):
...
     def _ad_annotate_apply(apply):
-        @no_annotations
         @wraps(apply)
         def wrapper(self, *args, **kwargs):
-            for arg in args:
-                if not hasattr(arg, "bcs"):
-                    arg.bcs = []
-            arg.bcs.append(self)
-            return apply(self, *args, **kwargs)
+            annotate = annotate_tape(kwargs)
```
Comment (pefarrell) [inline firedrake/adjoint/dirichletbc.py:25]: This is absolutely essential for defcon. Before this PR, defcon reliably crashed whenever I tried to run long bifurcation analyses, killed by the out-of-memory manager. After this change, the same code runs happily for weeks. Thanks so much, @wence- !

**Review by pefarrell — APPROVED** [review]

File: .gitignore (hunk 1 above)
Comment (wence-) [reply .gitignore:12]: I guess I could spin it out: it's cos I was adding the firedrake.adjoint.rst ignore. (This is a file MacOS finder dumps everywhere).

Comment (pefarrell) [reply .gitignore:12]: Not worth the effort, keep it as-is.

**Pull request merged**